internal/repository/repository.go

```diff
@@ -767,43 +768,89 @@ type Loader interface {
 	Load(ctx context.Context, h restic.Handle, length int, offset int64, fn func(rd io.Reader) error) error
 }
 
-// DownloadAndHash is all-in-one helper to download content of the file at h to a temporary filesystem location
-// and calculate ID of the contents. Returned (temporary) file is positioned at the beginning of the file;
-// it is the reponsibility of the caller to close and delete the file.
-func DownloadAndHash(ctx context.Context, be Loader, h restic.Handle) (tmpfile *os.File, hash restic.ID, size int64, err error) {
-	tmpfile, err = fs.TempFile("", "restic-temp-")
-	if err != nil {
-		return nil, restic.ID{}, -1, errors.Wrap(err, "TempFile")
+type BackendLoadFn func(ctx context.Context, h restic.Handle, length int, offset int64, fn func(rd io.Reader) error) error
+
+func StreamPack(ctx context.Context, beLoad BackendLoadFn, key *crypto.Key, packID restic.ID, blobs []restic.Blob, handleBlobFn func(blob restic.BlobHandle, buf []byte, err error) error) error {
```

Comment:
Why not use goroutines instead of a callback? Like,

```go
type BlobContent struct {
	restic.BlobHandle
	Plaintext []byte
	Err       error
}

func StreamPack(ctx context.Context, beLoad BackendLoadFn, key *crypto.Key, packID restic.ID, blobs []restic.Blob) <-chan *BlobContent
```

Reply:
The handleBlobFn can currently return an error to force StreamPack to try loading the blobs again, which won't work with goroutines. My idea here was to allow retrying broken downloads or temporary decryption errors. That functionality would currently only be used in `Repack` in the first `if err != nil`. However, I've just noticed that the error handling for `SaveBlob` in `Repack` was broken. As that error is one which cannot be retried.

So I guess, I'll remove at least the return value of `handleBlobFn`. The downside of letting StreamPack return a channel is that we also should make sure that the goroutine is shut down properly, which would probably also require using an ErrGroup as for `StreamTrees`.

Reply:
Hmm, looks like I can't get rid of the return value of `handleBlobFn`. If the filerestorer or repack encounters an error then `StreamPack` should just abort with an error. I've changed `StreamPack` to just interpret any error returned by the `handleBlobFn` as a permanent error, to prevent retries. It would be possible to get something similar by using errgroups, but in a quick test I ended up with a really messy error handling in the filerestorer.

Another complication of returning a channel would be that the plaintext buffer can no longer be reused easily.